### PR TITLE
Use RuntimeError in abort() before module instantiation

### PR DESCRIPTION
**Pull request by aheejin**
We decided use a trap for `abort` function in case of Wasm EH in order to prevent infinite-looping (#16910).
https://github.com/emscripten-core/emscripten/blob/44b2c2a1ecad39c534e14179acb49419dfee528b/src/preamble.js#L472-L474

The short reason is, in Wasm EH `RuntimeError` is treated as a foreign exception and caught by `catch_all`, so you try to abort the program and throw a `RuntimeError`, but it is caught by some `catch_all` used in the cleanup (=destructors, etc) code, causing the program to continue to run.

`__trap` is [defined](https://github.com/emscripten-core/emscripten/blob/44b2c2a1ecad39c534e14179acb49419dfee528b/system/lib/compiler-rt/__trap.c#L1-L3) in compiler-rt and [exported](https://github.com/emscripten-core/emscripten/blob/44b2c2a1ecad39c534e14179acb49419dfee528b/emcc.py#L2798-L2799) when Wasm EH is enabled.

This has worked so far, but in case we fail to instantiate a Wasm module and call `abort` because of it, we don't have access to the imported `Module['asm']['__trap']`, because `Module['asm']` has not been set:
https://github.com/emscripten-core/emscripten/blob/44b2c2a1ecad39c534e14179acb49419dfee528b/src/preamble.js#L848

And we haven't had the opportunity to run this code:
https://github.com/emscripten-core/emscripten/blob/44b2c2a1ecad39c534e14179acb49419dfee528b/src/preamble.js#L975

So the `abort` call will end like this:
```console
TypeError: Cannot read properties of undefined (reading '__trap')
    at ___trap (/usr/local/google/home/aheejin/test/gl84/exported_api.js:5152:34)
    at abort (/usr/local/google/home/aheejin/test/gl84/exported_api.js:892:5)
    at /usr/local/google/home/aheejin/test/gl84/exported_api.js:1172:5
```
which may not be the worst thing in the world given that we are crashing anyway, but not the situation we intended.

This PR lets us throw `RuntimeError` in case we don't have the wasm module instantiated and have access to `Module['asm']['__trap']`. This is OK even with Wasm EH because we haven't been running the module, there's no risk of infinite-looping we tried to prevent in #16910.

I'm not sure how to add a test for this, because the test would require a module that fails to instantiate. This was found while I was debugging something else.

### CODE REVIEW

File: src/preamble.js [hunk 1]
```diff
@@ -469,20 +469,37 @@ function abort(what) {
...
+  // Wasm EH code (because RuntimeError is considered as a foreign exception and
+  // caught by 'catch_all'), but in case throwing RuntimeError is fine because
+  // the module has not even been instantiated, even less running.
+  if (typeof Module['asm'] !== 'undefined')
+    ___trap();
+  else
+    throwError(what);
+#else
+  throwError(what);
 #endif
```
Comment (kripken) [inline src/preamble.js:503]:
```suggestion
  if (typeof Module['asm'] !== 'undefined') {
    ___trap();
  }
#endif
  throwError(what);
```

Comment (sbc100) [reply src/preamble.js:503]: Ineeded, and perhaps this change can avoid defining a new inner `throwError` function at all?   Can `__trap` just act as an early return followed by the rest of the throwError body?

**Comment by sbc100** [comment]
Do you know where the offending call to abort is coming from exactly?  I'm happy to land this fix but it also seems wrong to be a calling abort if the program isn't even running yet.  I'd like to look into fixing that too.

File: src/preamble.js [hunk 2]
```diff
@@ -469,20 +469,37 @@ function abort(what) {
...
+  //
+  // In case abort() is called before the module is initialized, Module['asm']
+  // and its exported '__trap' function is not available, in which case we throw
+  // a RuntimeError.
+  //
+  // We trap instead of throwing RuntimeError to prevent infinite-looping in
+  // Wasm EH code (because RuntimeError is considered as a foreign exception and
+  // caught by 'catch_all'), but in case throwing RuntimeError is fine because
+  // the module has not even been instantiated, even less running.
+  if (typeof Module['asm'] !== 'undefined')
```
[the remark this answers is not in the document]
Comment (aheejin) [reply src/preamble.js:497]: `if (runtimeInitialized)` seems to work, thanks. Didn't know that variable exited.

File: src/preamble.js (hunk 1 above)
Comment (aheejin) [reply src/preamble.js]: Yeah `__trap()` is an early return and it looks we don't need `throwError` after all. Fixed.

**Review by kripken — APPROVED** [review]
lgtm but I'm also curious where the problem happens. Is it in a global ctor or other startup code?

**Comment by kripken** [comment]
Oh, I think it can't be the wasm module at all, because then `runtimeInitialized` would be true..?

Then I think it has to be what you said before @aheejin , that it errors during a failure to instantiate the wasm in the JS code. @sbc100 I think such a call to `abort()` is reasonable? I'm not sure what we'd replace it with.

**Comment by aheejin** [comment]
The case I was debugging was this: https://github.com/emscripten-core/emscripten/issues/18592#issuecomment-1415863762

And the callsite involved in this case was here, in `instantiateArrayBuffer` (click to see the code around the line): https://github.com/emscripten-core/emscripten/blob/30ba0ae216857d29e7af8a581377b6beb5938d30/src/preamble.js#L848

Actually, it is not the initialization itself that fails, even though this can happen when intitialization fails too. What fails in this example is this line: https://github.com/emscripten-core/emscripten/blob/30ba0ae216857d29e7af8a581377b6beb5938d30/src/preamble.js#L818
which transfers the control flow here: https://github.com/emscripten-core/emscripten/blob/30ba0ae216857d29e7af8a581377b6beb5938d30/src/preamble.js#L821-L849
which ends with the `abort` call.

**Review by sbc100 — APPROVED** [review]

**Comment by kripken** [comment]
Makes sense, thanks for the details @aheejin

**Pull request merged**